Rocket.Chat/Controllers/Subscriptions/FilesListViewController.swift

```diff
@@ -0,0 +1,319 @@
+            }
+            present(ImageViewerController(configuration: configuration), animated: true)
+        } else {
+            openRemoteImage(fromFile: file, fromImageView: imageView)
+        }
+    }
+
+    private func openRemoteImage(fromFile file: File, fromImageView imageView: FLAnimatedImageView?) {
+        guard let fileURL = file.fullFileURL() else { return }
+
+        func open(data: Data?) {
+            let configuration = ImageViewerConfiguration { config in
+                if let data = data {
+                    if file.isGif {
+                        config.animatedImage = FLAnimatedImage(gifData: data)
+                    } else {
+                        config.image = UIImage(data: data)
+                    }
+                }
+                config.imageView = imageView
+            }
+
+            DispatchQueue.main.async {
+                MBProgressHUD.hide(for: self.view, animated: true)
+                self.present(ImageViewerController(configuration: configuration), animated: false)
+            }
+        }
+
+        MBProgressHUD.showAdded(to: view, animated: true)
+        SDWebImageDownloader.shared().downloadImage(with: fileURL, options: [.useNSURLCache], progress: nil, completed: { _, data, _, _ in
+            open(data: data)
+        })
+    }
+
+    func openVideo(fromFile file: File) {
+        guard let videoURL = file.fullFileURL() else { return }
+        let controller = MobilePlayerViewController(contentURL: videoURL)
+        controller.title = file.name
+        controller.activityItems = [file.name, videoURL]
+        present(controller, animated: true, completion: nil)
+    }
+
+    func openDocument(fromFile file: File) {
+        guard let identifier = file.identifier else { return }
+        guard let fileURL = file.fullFileURL() else { return }
+        let localUniqueName = identifier + file.url.replacingOccurrences(of: "/", with: "")
+        guard let filename = DownloadManager.filenameFor(localUniqueName) else { return }
+        guard let localFileURL = DownloadManager.localFileURLFor(filename) else { return }
+
+        // Open document itself
+        func open() {
+            documentController = UIDocumentInteractionController(url: localFileURL)
+            documentController?.delegate = self
+            DispatchQueue.main.async {
+                self.documentController?.presentPreview(animated: true)
+            }
+        }
+
+        // Checks if we do have the file in the system, before downloading it
+        if DownloadManager.fileExists(localFileURL) {
+            open()
+        } else {
+            let message = String(format: localized("chat.download.downloading_file"), file.name)
+            let loadingHUD = MBProgressHUD.showAdded(to: view, animated: true)
+            loadingHUD.label.text = message
+            // Download file and cache it to be used later
+            DownloadManager.download(url: fileURL, to: localFileURL) {
```

Comment:
[weak self]?

Reply:
Honestly I don't know how it behave when we have a reference for `self` inside a sub function that is called inside an async-operation callback 🤔 I will update this sub function to a closure then capture the weak reference of self to make sure it will not cause any problem.